server/src/transport/ws.rs

```diff
@@ -304,112 +304,42 @@ pub(crate) async fn background_task<L: Logger>(
-							}
-
-							CallOrSubscription::Call(r) => {
-								logger.on_response(&r.result, request_start, TransportProtocol::WebSocket);
-								sink_permit.send_raw(r.result);
-							}
-						}
-					};
-				}
-				.boxed();
-
-				method_executor.add(fut);
-			}
-			Some(b'[') => {
-				let limit = match batch_requests_config {
-					BatchRequestConfig::Disabled => {
-						let response = MethodResponse::error(
-							Id::Null,
-							ErrorObject::borrowed(BATCHES_NOT_SUPPORTED_CODE, &BATCHES_NOT_SUPPORTED_MSG, None),
-						);
-						logger.on_response(&response.result, request_start, TransportProtocol::WebSocket);
-						sink_permit.send_raw(response.result);
-						continue;
-					}
-					BatchRequestConfig::Limit(limit) => limit as usize,
-					BatchRequestConfig::Unlimited => usize::MAX,
-				};
+		};
 
-				// Make sure the following variables are not moved into async closure below.
-				let methods = &methods;
-				let sink = sink.clone();
-				let id_provider = id_provider.clone();
-				let data = std::mem::take(&mut data);
-				let bounded_subscriptions = bounded_subscriptions.clone();
-
-				let fut = async move {
-					let response = process_batch_request(Batch {
-						data,
-						call: CallData {
-							conn_id: conn_id as usize,
-							bounded_subscriptions,
-							max_response_body_size,
-							max_log_length,
-							methods,
-							sink: &sink,
-							id_provider: &*id_provider,
-							logger,
-							request_start,
-						},
-						max_len: limit,
-					})
-					.await;
-
-					if let Some(response) = response {
-						tx_log_from_str(&response, max_log_length);
-						logger.on_response(&response, request_start, TransportProtocol::WebSocket);
-						sink_permit.send_raw(response);
-					}
-				};
+		let params = ExecuteCallParams {
+			batch_requests_config,
+			bounded_subscriptions: bounded_subscriptions.clone(),
+			conn_id,
+			methods: methods.clone(),
+			max_log_length,
+			max_response_body_size,
+			sink: sink.clone(),
+			sink_permit,
+			id_provider: id_provider.clone(),
+			logger: logger.clone(),
+			data: std::mem::take(&mut data),
+		};
 
-				method_executor.add(Box::pin(fut));
-			}
-			_ => {
-				sink_permit.send_error(Id::Null, ErrorCode::ParseError.into());
-			}
-		}
+		pending_calls.push(tokio::spawn(execute_unchecked_call(params)));
 	};
 
-	logger.on_disconnect(remote_addr, TransportProtocol::WebSocket);
-
 	// Drive all running methods to completion.
 	// **NOTE** Do not return early in this function. This `await` needs to run to guarantee
 	// proper drop behaviour.
-	method_executor.await;
+	//
+	// This is not strictly not needed because `tokio::spawn` will drive these the completion
+	// but it's preferred that the `stop_handle.stopped()` should not return until all methods has been
+	// executed and the connection has been closed.
+	tokio::select! {
+		// All pending calls executed.
+		_ = pending_calls.for_each(|_| async {}) => {
```

Comment:
So, we could also timeout out this after 10 minutes or something if someone has a infinite loop or something that never returns in some scenario it may "leak" and never terminate...